theano/compile/debugmode.py

```diff
@@ -2293,7 +2298,7 @@ def __init__(self, inputs, outputs, optimizer, mode,
                 for output, spec in izip(fgraph.outputs,
                                          outputs + additional_outputs)
                 if not spec.borrow]
-        if no_borrow:
+        if no_borrow and not always_borrow:
```

Comment:
I checked the code, and only the usage of In do not work with shared variable. The usage of Out with shared variable should work.